presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java

```diff
@@ -478,25 +488,105 @@ private PlanBuilder appendSubqueryApplyNode(PlanBuilder subPlan, Expression subq
+                        orElse(ImmutableList.of()));
+    }
+
+    private PlanNode projectAllCoercionsIfNeededWithTranslations(TranslationMap translations, PlanNode subqueryPlanNode, Expression subqueryExpression)
+    {
+        return projectCoercionsIfNeeded(translations, subqueryPlanNode, subqueryExpression, true, extractCoercions(subqueryExpression));
+    }
+
+    private PlanNode projectCoercionsIfNeeded(TranslationMap translations, PlanNode subqueryPlanNode, Expression subqueryExpression, boolean addTranslations, List<Coercion> coercions)
+    {
+        if (subqueryPlanNode.getOutputSymbols().size() != 1) {
+            return subqueryPlanNode;
+        }
+
+        Symbol subquerySymbol = getOnlyElement(subqueryPlanNode.getOutputSymbols());
+        if (coercions.isEmpty()) {
+            // subquery symbol is only used in direct context
+            if (addTranslations) {
+                translations.put(subqueryExpression, subquerySymbol);
+            }
+            return subqueryPlanNode;
+        }
+
+        // for each subqueryExpression coercion add a casting projection
+        Type subqueryType = analysis.getType(subqueryExpression);
+        ImmutableMap.Builder<Symbol, Expression> assignments = ImmutableMap.builder();
+        coercions.stream()
+                .forEach(coercion -> {
+                    Symbol coercionSymbol = symbolAllocator.newSymbol(subqueryExpression, coercion.getType());
+                    Expression castExpression =
+                            new Cast(
+                                    subquerySymbol.toSymbolReference(),
+                                    coercion.getType().getTypeSignature().toString(),
+                                    false,
+                                    metadata.getTypeManager().isTypeOnlyCoercion(subqueryType, coercion.getType()));
+                    assignments.put(coercionSymbol, castExpression);
+                    if (addTranslations) {
+                        translations.put(coercion.getExpression(), coercionSymbol);
+                    }
+                });
+
+        if (!analysis.getCoercions().containsKey(subqueryExpression)) {
+            // subquery symbol is also used in direct context
+            assignments.put(subquerySymbol, subquerySymbol.toSymbolReference());
+            if (addTranslations) {
+                translations.put(subqueryExpression, subquerySymbol);
+            }
+        }
+
+        return new ProjectNode(
+                idAllocator.getNextId(),
+                subqueryPlanNode,
+                assignments.build());
+    }
+
+    private Optional<Coercion> extractExactCoercion(Expression subqueryExpression)
+    {
+        return analysis.getCoercions().entrySet().stream()
+                .filter(entry -> entry.getKey() == subqueryExpression)
+                .map(entry -> new Coercion(entry.getKey(), entry.getValue()))
+                .findFirst();
```

Comment:
it should be safe to collect to a list and then check if there is at most single element